Add api_fields configuration value to specify which fields to return

robbieaverill opened the pull request:
* Add tests
* Update docs

colymba commented:
Looks good. Can we extend this to relations as well? So `api_fields` would be for `$db` `$has_one` `$has_many` etc...

robbieaverill commented:
Hey Thierry, isn't it recursive? I didn't test it but I'd hoped it would
work for nested relations too

On Monday, 27 June 2016, Thierry François <notifications@github.com> wrote:

> Looks good. Can we extend this to relations as well? So api_fields would
> be for $db $has_one $has_many etc...
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/colymba/silverstripe-restfulapi/pull/70#issuecomment-228671576>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AE7lnrCryo8420tQXAh4w3b18Aw0HqUYks5qP3qQgaJpZM4I-dW1>
> .
>


-- 
*Robbie Averill*
LA PETITE MANOUCHE - NZ Gypsy Jazz Guitar Duo
Phone: +64 27 321 4758
contact@lapetitemanouche.com
www.lapetitemanouche.com

colymba commented:
Not sure we are talking about the same thing 😄  but yes as it is it would work for any record, the main one or embedded ones from relations but only on `db` fields.

But for now the `has_one` and `_many` relations will still be included as ID or embedded depending on config.

Adding the same check you did on lines https://github.com/robbieaverill/silverstripe-restfulapi/blob/0def6112c4501a77412254b5b65a7d9bd9a8eabc/code/serializers/Basic/RESTfulAPI_BasicSerializer.php#L169 and https://github.com/robbieaverill/silverstripe-restfulapi/blob/0def6112c4501a77412254b5b65a7d9bd9a8eabc/code/serializers/Basic/RESTfulAPI_BasicSerializer.php#L206 would solve this.

robbieaverill commented:
@colymba yes you're totally right. I naively assumed it was calling `formatDataObject` on each of the relation models too. I'll update now for the other relations.

robbieaverill commented:
Hey @colymba 

I've updated the code and the tests and docs.

I'm very sorry, but I got sick of my linter showing me entire screens of red so I've fixed the indentation in the files I was working in :) If you want me to revert that let me know, happy to undo it.

Have a test and let me know what you think.

Cheers,
Robbie

colymba commented on code/serializers/Basic/RESTfulAPI_BasicSerializer.php:
```diff
@@ -116,82 +112,94 @@ public function serialize($data)
-        foreach ($has_one as $columnName => $fieldClassName) {
+        // Get a possibly defined list of "api_fields" for this DataObject. If defined, they will be the only fields
+        // for this DataObject that will be returned, including related models.
+        $apiFields = (array) Config::inst()->get($dataObject->ClassName, 'api_fields');
```
Could we have the `api_fields` declaration below the other `db`, `has_one`, `has_many` etc on line 140? Just easier to follow the code and what `api_fields` relates to....

colymba commented:
Indentation is fine... the whole module could use an overall anyway...
Just posted one last comment before merging!

Thanks @robbieaverill

robbieaverill commented:
Sure, all yours @colymba - thanks

colymba commented:
thanks @robbieaverill, sorry last request, can it be squashed into 1 commit too?

robbieaverill commented:
Of course you can!

colymba commented:
Awesome! thanks @robbieaverill
